File: src/bmsx/runtime/assets/edits.ts
import { $ } from '../../core/engine';
import { taskGate } from '../../core/taskgate';
import type { Memory } from '../../machine/memory/memory';

const runtimeAssetEditGate = taskGate.group('asset:update');

export function flushRuntimeAssetEdits(memory: Memory): void {
	const dirty = memory.consumeDirtyAssets();
	if (dirty.length === 0) {
		return;
	}
	for (let index = 0; index < dirty.length; index += 1) {
		const entry = dirty.get(index);
		if (entry.type === 'image') {
			const vramSpan = entry.capacity > 0 ? entry.capacity : 1;
			if (memory.isVramRange(entry.baseAddr, vramSpan)) {
				continue;
			}
			const token = runtimeAssetEditGate.begin({ blocking: false, category: 'texture', tag: `asset:${entry.id}` });
			void $.texmanager.updateTexturesForKey(entry.id, memory.getImagePixels(entry), entry.regionW, entry.regionH)
				.finally(() => runtimeAssetEditGate.end(token));
		} else if (entry.type === 'audio') {
			$.sndmaster.invalidateClip(entry.id);
		}
	}
}

File: src/bmsx_cpp/runtime/assets/edits.cpp
#include "runtime/assets/edits.h"

#include "core/engine.h"
#include "core/primitives.h"
#include "machine/memory/memory.h"
#include "render/gameview.h"
#include "render/texture_manager.h"

namespace bmsx {

void flushRuntimeAssetEdits(Memory& memory) {
	auto* view = EngineCore::instance().view();
	if (!view->backend()->readyForTextureUpload()) {
		return;
	}
	auto dirty = memory.consumeDirtyAssets();
	if (dirty.empty()) {
		return;
	}
	auto* texmanager = EngineCore::instance().texmanager();
	for (const auto* entry : dirty) {
		if (entry->type != Memory::AssetType::Image) {
			continue;
		}
		const uint32_t span = entry->capacity > 0 ? entry->capacity : 1u;
		if (memory.isVramRange(entry->baseAddr, span)) {
			continue;
		}
		texmanager->updateTexturesForAsset(
			entry->id,
			memory.getImagePixels(*entry),
			static_cast<i32>(entry->regionW),
			static_cast<i32>(entry->regionH)
		);
	}
}

} // namespace bmsx
